python/pyarrow/_flight.pyx

```diff
@@ -1667,6 +1812,258 @@ cdef class ClientAuthHandler:
+            If an exception is raised, the call will be rejected with
+            the given error.
+
+        """
+
+
+cdef class ServerMiddleware:
+    """Server-side middleware for a call, instantiated per RPC.
+
+    Methods here should be fast and must be infalliable: they should
+    not raise exceptions or stall indefinitely.
+
+    """
+
+    def sending_headers(self):
+        """A callback before headers are sent.
+
+        Returns
+        -------
+        headers : dict
+            A dictionary of header values to add to the response, or
+            None if no headers are to be added. The dictionary should
+            have string keys and string or list-of-string values.
+
+        """
+
+    def call_completed(self, exception):
+        """A callback when the call finishes.
+
+        Parameters
+        ----------
+        exception : pyarrow.ArrowException
+            If the call errored, this is the equivalent
+            exception. Will be None if the call succeeded.
+
+        """
+
+    @staticmethod
+    cdef void wrap(object py_middleware,
+                   shared_ptr[CServerMiddleware]* c_instance):
+        cdef PyServerMiddlewareVtable vtable
+        vtable.sending_headers = _middleware_sending_headers
+        vtable.call_completed = _middleware_call_completed
+        c_instance[0].reset(new CPyServerMiddleware(py_middleware, vtable))
+
+
+cdef class _ServerMiddlewareFactoryWrapper(ServerMiddlewareFactory):
+    """A wrapper that bundles server middleware into a single C++ one.
+
+    gRPC runs RPCs on C++ threads, so we use PyGilState_Ensure/Release
+    to interact with the GIL before running Python code. However, this
+    has a major shortcoming: it breaks thread locals. When we enter
+    Python, it creates new thread state. When we exit, that thread
+    state is freed (and Python thread locals cleaned up), since the
+    thread isn't persistent from Python's point of view. This makes
+    development of middleware harder than it needs to be.
+
+    This middleware solves this by first incrementing the reference
+    count on the thread state, then running all Python middleware. It
+    then doesn't decrement the reference count until the call is
+    finished.
+
+    """
+
+    cdef:
+        list factories
+
+    def __init__(self, factories):
+        self.factories = factories
+
+    def start_call(self, info, headers):
+        cdef PyGILState_STATE state = PyGILState_Ensure()
+
+        instances = []
+        try:
+            for factory in self.factories:
+                instance = factory.start_call(info, headers)
+                if instance:
+                    instances.append(instance)
+        except:  # noqa: E722
+            # Don't artifically keep thread state around if an error
+            # happens.
+            # We want a bare except since we always want to release
+            # thread state.
+            PyGILState_Release(state)
+            raise
```

Comment:
Raising a Python exception with the GIL released is completely undefined.

Reply:
This won't actually release the GIL, just the local reference to the thread state, because the C++ side is also holding the GIL.

Reply:
Then it shouldn't be necessary at all...

Reply:
The intent of this (which I guess is moot now) is that in the non-exception path, the GIL would remain held by Cython while C++ would release its reference, so that Python wouldn't clean up thread state until the request finished. Otherwise, we keep jumping in and out of Python, destroying thread state every time.

In retrospect, that would mean that Python server performance would plummet, because previously most of the C++ plumbing did not need the GIL. We don't really actually want the GIL here, what we wanted was to preserve the Python thread state until we're done.

I should probably follow with a better set of Flight benchmarks across all languages...